autoprecompiles/src/optimizer.rs

```diff
@@ -0,0 +1,182 @@
+    qse_opt::{
+        algebraic_to_quadratic_symbolic_expression, quadratic_symbolic_expression_to_algebraic,
+        Variable,
+    },
+    simplify_expression,
+};
+
+use crate::{BusInteractionKind, SymbolicBusInteraction, SymbolicConstraint, SymbolicMachine};
+
+/// Simplifies the constraints as much as possible.
+/// This function is similar to powdr_pilopt::qse_opt::run_qse_optimization, except it:
+/// - Runs on the entire constraint system, including bus interactions.
+/// - Panics if the solver fails.
+/// - Removes trivial constraints (e.g. `0 = 0` or bus interaction with multiplicity `0`)
+///   from the constraint system.
+/// - Calls `simplify_expression()` on the resulting expressions.
+pub fn optimize<P: FieldElement>(symbolic_machine: SymbolicMachine<P>) -> SymbolicMachine<P> {
+    let constraint_system = symbolic_machine_to_constraint_system(symbolic_machine);
+
+    log_constraint_system_stats("Starting optimize()", &constraint_system);
+    let constraint_system = solver_based_optimization(constraint_system);
+    log_constraint_system_stats("After solver-based optimization", &constraint_system);
+    let constraint_system = remove_trivial_constraints(constraint_system);
+    log_constraint_system_stats("After removing trivial constraints", &constraint_system);
+
+    // TODO: Add equivalent of replace_linear_witness_columns step to make
+    // powdr_optimize_legacy obsolete
+
+    constraint_system_to_symbolic_machine(constraint_system)
+}
+
+fn symbolic_machine_to_constraint_system<P: FieldElement>(
+    symbolic_machine: SymbolicMachine<P>,
+) -> ConstraintSystem<P, Variable> {
+    ConstraintSystem {
+        algebraic_constraints: symbolic_machine
+            .constraints
+            .iter()
+            .map(|constraint| algebraic_to_quadratic_symbolic_expression(&constraint.expr))
+            .collect(),
+        bus_interactions: symbolic_machine
+            .bus_interactions
+            .iter()
+            .map(symbolic_bus_interaction_to_bus_interaction)
+            .collect(),
+    }
+}
+
+fn constraint_system_to_symbolic_machine<P: FieldElement>(
+    constraint_system: ConstraintSystem<P, Variable>,
+) -> SymbolicMachine<P> {
+    SymbolicMachine {
+        constraints: constraint_system
+            .algebraic_constraints
+            .iter()
+            .map(|constraint| SymbolicConstraint {
+                expr: simplify_expression(quadratic_symbolic_expression_to_algebraic(constraint)),
+            })
+            .collect(),
+        bus_interactions: constraint_system
+            .bus_interactions
+            .into_iter()
+            .map(bus_interaction_to_symbolic_bus_interaction)
+            .collect(),
+    }
+}
+
+fn solver_based_optimization<T: FieldElement>(
+    constraint_system: ConstraintSystem<T, Variable>,
+) -> ConstraintSystem<T, Variable> {
+    let result = Solver::new(constraint_system)
+        .solve()
+        .map_err(|e| {
+            panic!("Solver failed: {e:?}");
```

Comment:
what does it mean for the solver to fail?

Reply:
It means that the constraints are contradictory, which would be a bug in the APC code.